### PR TITLE
add null check for module-default-defs

### PR DESCRIPTION
**Pull request by ssikdar1**
# Description

Issue #34544.
In PR #35337  there was a suggestion to add a check for `body` in `module-default-defs` in `src/jlfrontend.scm`. This adds that check.

# Testing
In the julia repl I tried a couple more variations that would cause a segfault
```
julia> eval(Expr(:call, :eval, Expr(:quote, Expr(:module, true, :bar, Expr(:block)))))
Main.bar

julia> eval(Expr(:module, true, :bar, Expr(:block)))
WARNING: replacing module bar.
Main.bar

julia> eval(Expr(:module, true, :bar, Expr(:quote)))
WARNING: replacing module bar.
Main.bar

julia> eval(Expr(:call, :eval, Expr(:quote, Expr(:module, true, :bar, Expr(:block)))))
WARNING: replacing module bar.
Main.bar
```

### CODE REVIEW

**Comment by JeffBezanson** [comment]
Great, thanks! Could you add a test case to test/syntax.jl?

**Comment by ssikdar1** [comment]
5a0b565 adds a fix for #35367 as well

Output for those two lines in the REPL:
```
$ ./julia 
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.5.0-DEV.572 (2020-04-06)
 _/ |\__'_|_|_|\__'_|  |  linenumber/0b9ed994df* (fork: 2 commits, 4 days)
|__/                   |

julia> eval(Expr(:quote, Expr(:module, true, :bar, Expr(:quote))))
:(module bar
  end)

julia> eval(Expr(:call, :eval, Expr(:quote, Expr(:module, true, :bar, Expr(:quote)))))
Main.bar
```

File: src/ast.c [hunk 1]
```diff
@@ -575,7 +575,7 @@ static jl_value_t *scm_to_julia_(fl_context_t *fl_ctx, value_t e, jl_module_t *m
             assert(jl_is_symbol(ex));
             temp = jl_module_globalref(jl_core_module, (jl_sym_t*)ex);
         }
-        else if (sym == inert_sym || (sym == quote_sym && (!iscons(car_(e))))) {
+        else if ((sym == inert_sym || (sym == quote_sym && (!iscons(car_(e))))) && iscons(e)) {
```
Comment (JeffBezanson) [inline src/ast.c:578]: `iscons(e)` needs to be checked before `car_(e)`; `car` is only valid if `e` is a cons.

File: test/core.jl [hunk 2]
```diff
@@ -7198,3 +7198,11 @@ struct NFANode34126
 end
 
 @test repr(NFANode34126()) == "$NFANode34126(Tuple{Nothing,$NFANode34126}[])"
+
+@testset "issue #34544/35367" begin
+    # Test these evals dont throw an error
+    eval(Expr(:call, :eval, Expr(:quote, Expr(:module, true, :bar, Expr(:block)))))
+    eval(Expr(:module, true, :bar, Expr(:block)))
+    eval(Expr(:quote, Expr(:module, true, :bar, Expr(:quote))))
+    eval(Expr(:call, :eval, Expr(:quote, Expr(:module, true, :bar, Expr(:quote)))))
```
Comment (JeffBezanson) [inline test/core.jl:7207]: I don't think this should be valid; the 3rd argument to `module` has to be a `block` expression. That can be checked at the top of `jl_eval_module_expr`.

**Comment by ssikdar1** [comment]
@JeffBezanson added an if check  inside `jl_eval_module_expr` in dce7897.

The following now throw an Exception:
```
julia> eval(Expr(:call, :eval, Expr(:quote, Expr(:module, true, :bar, Expr(:quote)))))
ERROR: syntax: module expression third argument must be a block

julia> eval(Expr(:module, true, :bar, Expr(:foo)))
ERROR: syntax: module expression third argument must be a block

julia> eval(Expr(:module, true, :bar, Expr(:quote)))
ERROR: syntax: module expression third argument must be a block
```

The following does not return an ErrorException (but no longer segfaults ):
```
julia> eval(Expr(:quote, Expr(:module, true, :bar, Expr(:quote))))
:(module bar
  end)
```
I think due to the quoting? Which I think is fine?

**Comment by mbauman** [comment]
I just hit this bug myself — glad to see a fix already here!  My use-case was a macro:

```Julia
julia> macro foo()
           Expr(:module, true, esc(:Bar), Expr(:block))
       end
@foo (macro with 1 method)

julia> @foo
WARNING: replacing module Bar.

signal (11): Segmentation fault: 11
```

**Comment by JeffBezanson** [comment]
> I think due to the quoting? Which I think is fine?

Right; the `eval` just strips off the quote, so the module expression itself is never evaluated, so the error doesn't happen.

**Pull request merged**